### PR TITLE
Object Detection with Gaudi2

### PR DESCRIPTION
**Pull request by kprokofi**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by sungchul1** [comment]
@kprokofi I and @eunwoosh opened https://github.com/openvinotoolkit/training_extensions/pull/2574, could you update this PR after merging https://github.com/openvinotoolkit/training_extensions/pull/2574?

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py [hunk 1]
```diff
@@ -198,6 +223,18 @@ def train_detector(model, dataset, cfg, distributed=False, validate=False, times
...
+        return cfg_optim
+
+    # Only SGD for OD and IS supported by now on HPU
+    cfg_optim["type"] = "SGD"
+    if "betas" in cfg_optim:
+        del cfg_optim["betas"]
+    return cfg_optim
+
+
 def monkey_patched_xpu_nms(ctx, bboxes, scores, iou_threshold, offset, score_threshold, max_num):
```
Comment (sungchul1) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: I think `xpu` in the function name can be deleted.

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py [hunk 2]
```diff
@@ -137,6 +154,14 @@ def train_detector(model, dataset, cfg, distributed=False, validate=False, times
         model.train()
         model, optimizer = torch.xpu.optimize(model, optimizer=optimizer, dtype=dtype)
 
+    if cfg.device == "hpu":
+        NMSop.forward = monkey_patched_xpu_nms
+        RoIAlign.forward = monkey_patched_xpu_roi_align
```
Comment (sungchul1) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: `xpu` in these function names can be deleted.

File: src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py [hunk 3]
```diff
@@ -64,7 +64,7 @@ def build_data_parallel(
         model = model.xpu()
         model = XPUDataParallel(model, device_ids=config.gpu_ids)
     elif is_hpu_available() and config.get("gpu_ids", []):
-        model = model.hpu()
+        model = model.to("hpu")
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py:67]: It's just a question. Are they different?

File: src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py [hunk 4]
```diff
@@ -153,6 +154,8 @@ def scatter(self, inputs, kwargs, device_ids):
                 for val in x:
                     if isinstance(val, dict):
                         for k in val:
+                            if not self.put_gt_on_device and k.startswith("gt_"):
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py:159]: Could you add a comment explaning what it's for?

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py [hunk 5]
```diff
@@ -119,11 +121,26 @@ def train_detector(model, dataset, cfg, distributed=False, validate=False, times
     elif cfg.device == "xpu":
         model = build_dp(model, cfg.device, device_ids=cfg.gpu_ids, enable_autocast=bool(fp16_cfg))
         model.to(f"xpu:{cfg.gpu_ids[0]}")
+    elif cfg.device == "hpu":
+        import habana_frameworks.torch.core as htcore
+        from habana_frameworks.torch.utils.library_loader import load_habana_module
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: Could you remove code using `load_habana_module`? It isn't required according to https://docs.habana.ai/en/latest/Release_Notes/Release%20Notes%20v1.5.0.html?highlight=load_habana_module#id6

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py [hunk 6]
```diff
@@ -119,11 +121,26 @@ def train_detector(model, dataset, cfg, distributed=False, validate=False, times
...
+
+        load_habana_module()
+        os.environ["PT_HPU_LAZY_MODE"] = "1"
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: As far as I know, `PT_HPU_LAZY_MODE` is set as 1 in default setting. How about removing this code?

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py [hunk 7]
```diff
@@ -119,11 +121,26 @@ def train_detector(model, dataset, cfg, distributed=False, validate=False, times
...
+        assert len(cfg.gpu_ids) == 1
+        model = build_dp(
+            model, cfg.device, device_ids=cfg.gpu_ids, dim=0, enable_autocast=bool(fp16_cfg), put_gt_on_device=False
+        )
+        model.to(model.src_device_obj)
+        htcore.mark_step()
+        model.zero_grad()
```
Comment (sungchul1) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: Could you explain why they are here? AFAIK, they can be in train/eval loop.

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py [hunk 8]
```diff
@@ -119,11 +121,26 @@ def train_detector(model, dataset, cfg, distributed=False, validate=False, times
...
+        from habana_frameworks.torch.utils.library_loader import load_habana_module
+
+        load_habana_module()
+        os.environ["PT_HPU_LAZY_MODE"] = "1"
+        assert len(cfg.gpu_ids) == 1
+        model = build_dp(
+            model, cfg.device, device_ids=cfg.gpu_ids, dim=0, enable_autocast=bool(fp16_cfg), put_gt_on_device=False
+        )
+        model.to(model.src_device_obj)
+        htcore.mark_step()
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: It's just a question. Is it necessary for `htcore.mark_step()` to exist at this point?

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py (hunk 6 above)
Comment (sungchul1) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: It would be better to move here if possible.
https://github.com/openvinotoolkit/training_extensions/blob/575b4ec0740d1512af8826eb075085a4dbef0640/src/otx/algorithms/common/utils/__init__.py#L61-L62

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py [hunk 9]
```diff
@@ -119,11 +121,26 @@ def train_detector(model, dataset, cfg, distributed=False, validate=False, times
...
+        import habana_frameworks.torch.core as htcore
+        from habana_frameworks.torch.utils.library_loader import load_habana_module
+
+        load_habana_module()
+        os.environ["PT_HPU_LAZY_MODE"] = "1"
+        assert len(cfg.gpu_ids) == 1
+        model = build_dp(
+            model, cfg.device, device_ids=cfg.gpu_ids, dim=0, enable_autocast=bool(fp16_cfg), put_gt_on_device=False
+        )
+        model.to(model.src_device_obj)
```
Comment (sungchul1) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: @eunwoosh Does it seem same with segmentation? If not, `model` can be `model.cuda()` in `build_dp`.

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py [hunk 10]
```diff
@@ -137,6 +154,14 @@ def train_detector(model, dataset, cfg, distributed=False, validate=False, times
         model.train()
         model, optimizer = torch.xpu.optimize(model, optimizer=optimizer, dtype=dtype)
 
+    if cfg.device == "hpu":
+        NMSop.forward = monkey_patched_xpu_nms
+        RoIAlign.forward = monkey_patched_xpu_roi_align
+        # build runner
+        if cfg.device == "hpu":
+            if (new_type := "Fused" + cfg.optimizer.get("type", "SGD")) in HABANA_OPTIMIZERS:
+                cfg.optimizer["type"] = new_type
```
Comment (sungchul1) [inline src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: It should be set before building optimizer (L144).

File: src/otx/algorithms/detection/adapters/mmdet/evaluation/evaluator.py [hunk 11]
```diff
@@ -94,12 +100,19 @@ def print_map_summary(  # pylint: disable=too-many-locals,too-many-branches
...
-        table_data.append(["mAP", "", "", "", f"{mean_ap[i]:.3f}", f"{np.mean(mious[i]):.3f}"])
+        table_ = (
+            ["mAP", "", "", "", f"{mean_ap[i]:.3f}", f"{np.mean(mious[i]):.3f}"]
+            if segmentation
+            else ["mAP", "", "", "", f"{mean_ap[i]:.3f}"]
+        )
+        table_data.append(table_)
         table = AsciiTable(table_data)
         table.inner_footing_row_border = True
+        time.sleep(0.1)  # prevent segmentation fault
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/adapters/mmdet/evaluation/evaluator.py]: I think this code don't have to be excuted in a case where HPU device isn't used. How about change this code to be executed only in HPU case?

File: src/otx/algorithms/detection/adapters/mmdet/evaluation/evaluator.py [hunk 12]
```diff
@@ -59,6 +61,7 @@ def print_map_summary(  # pylint: disable=too-many-locals,too-many-branches
     if scale_ranges is not None:
         assert len(scale_ranges) == num_scales
 
+    segmentation = "miou" in results
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/adapters/mmdet/evaluation/evaluator.py:65]: It's just a question. Does `miou` make an error? What is this change for?

File: src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py [hunk 13]
```diff
@@ -153,6 +154,8 @@ def scatter(self, inputs, kwargs, device_ids):
...
+                                continue
```
Comment (sungchul1) [inline src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py:160]: It is just to use post processing on CPU, right?

**Review by sungchul1 — COMMENTED** [review]
I left comments. And I'm just curious if habana's OD models also use post-processing on CPU.

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py (hunk 9 above)
Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: Because `cfg.device` is `hpu`, `model.cuda()` isn't executed. So this code is necessary, I think.

Comment (sungchul1) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: I double checked in segmentation, with `model.cuda()` on gpu migration, model's weights are automatically converted to `hpu`.
`cfg.device` is not important in this situation because it is just for using `HPUDataParallel` in `dp_factory`.

```python
# before model.cuda()
for i, (name, params) in enumerate(model.named_parameters()):
    print(name, params.device)
    if i > 4: break

>>> backbone.stem.conv1.conv.weight cpu
>>> backbone.stem.conv1.bn.weight cpu
>>> backbone.stem.conv1.bn.bias cpu
>>> backbone.stem.branch1.0.conv.weight cpu
>>> backbone.stem.branch1.0.bn.weight cpu

model.cuda()
for i, (name, params) in enumerate(model.named_parameters()):
    print(name, params.device)
    if i > 4: break 

>>> backbone.stem.conv1.conv.weight hpu:0
>>> backbone.stem.conv1.bn.weight hpu:0
>>> backbone.stem.conv1.bn.bias hpu:0
>>> backbone.stem.branch1.0.conv.weight hpu:0
>>> backbone.stem.branch1.0.bn.weight hpu:0
>>> backbone.stem.branch1.0.bn.bias hpu:0
```

Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: Below code is `bulid_dp` function in mmseg. As you can see, because device is `hpu`, `model.cuda()` isn't executed. Due to same reason, `build_dp` in mmdet doesn't execute `model.cuda()` in `build_dp` function. This is mmdet [code](https://github.com/open-mmlab/mmdetection/ddddddddblob/v2.28.2/mmdet/utils/util_distribution.py#L10).
image

Comment (sungchul1) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: Yes, you're right. So, I suggest `model.cuda()` as `build_dp` args.
As you said, all if conditions will be skipped.
```python
model = build_dp(model.cuda(), cfg.device, ...) # cfg.device="hpu"
```

Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py:134]: I don't think so, because some of users want to use on CPU.

Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: Do you mean code like below?
```
elif cfg.device == "hpu":
    model = build_dp(
            model.cuda(), cfg.device, device_ids=cfg.gpu_ids, dim=0, enable_autocast=bool(fp16_cfg), put_gt_on_device=False
        )
```
Then it should work well. I think it's design choice. Do you think it's better than code below?
```
elif cfg.device == "hpu":
    model = build_dp(
            model, cfg.device, device_ids=cfg.gpu_ids, dim=0, enable_autocast=bool(fp16_cfg), put_gt_on_device=False
        )
    model.to(model.src_device_obj)
```

Comment (sungchul1) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: Oh, it seems that we don't need both of them because `DataParallel` casts model's device ([link](https://github.com/pytorch/pytorch/blob/e9ebda29d87ce0916ab08c06ab26fd3766a870e5/torch/nn/parallel/data_parallel.py#L147-L148)).
I checked `model` was converted to `hpu` after `model = build_dp(model, cfg.device, ...)`.
```python
# before model = build_dp(model, cfg.device, ...) in apis/train.py
for i, (name, params) in enumerate(model.named_parameters()):
    print(name, params.device)
    if i == 2: break
>>> backbone.stem.conv1.conv.weight cpu
>>> backbone.stem.conv1.bn.weight cpu
>>> backbone.stem.conv1.bn.bias cpu

# before super().__init__(*args, **kwargs) in _builder_build_data_parallel.py
for i, (name, params) in enumerate(args[0].named_parameters()):
    print(name, params.device)
    if i == 2: break
>>> backbone.stem.conv1.conv.weight cpu
>>> backbone.stem.conv1.bn.weight cpu
>>> backbone.stem.conv1.bn.bias cpu

super().__init__(*args, **kwargs)
for i, (name, params) in enumerate(args[0].named_parameters()):
    print(name, params.device)
    if i == 2: break
>>> backbone.stem.conv1.conv.weight hpu:0
>>> backbone.stem.conv1.bn.weight hpu:0
>>> backbone.stem.conv1.bn.bias hpu:0
```

Let's discuss it later because we need more tests, or escalate it to habana team.

Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: oh ok then we don't have to call explicitly. That's good news.

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py (hunk 7 above)
Comment (kprokofi) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: I just followed some code examples in Habana documentation. I don't think we need it either. I will remove this

File: src/otx/algorithms/detection/adapters/mmdet/evaluation/evaluator.py (hunk 12 above)
Comment (kprokofi) [reply src/otx/algorithms/detection/adapters/mmdet/evaluation/evaluator.py:65]: No, it is just needed to recognize instance segmentation task and object detection. If "miou" in results, then we deal with segmentation task.

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py (hunk 8 above)
Comment (kprokofi) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: I just followed documentation form Habana and they add this in tutorials

File: src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py (hunk 4 above)
Comment (kprokofi) [reply src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py:159]: Since post-processing is computed on CPU -> we don't need to put annotations on HPU

File: src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py (hunk 3 above)
Comment (kprokofi) [reply src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py:67]: model = model.hpu() throws an error

File: src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py (hunk 4 above)
Comment (eunwoosh) [reply src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py:159]: Ah I mean that could you add some comment nearby code to make us easier to understand code later.

File: src/otx/algorithms/detection/adapters/mmdet/apis/train.py (hunk 8 above)
Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/apis/train.py]: yes, I also saw each documentation has different `mark_step` usage. It seems that as Habana version increases, where it should be added is changed referring this [link](https://forum.habana.ai/t/when-to-use-htcode-mark-step/446/2).
I think `mark_step` after model initialization isn't necesaary. How about removing it?

File: src/otx/algorithms/detection/adapters/mmdet/evaluation/evaluator.py (hunk 12 above)
Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/evaluation/evaluator.py:65]: I see. It seems that some condition branch depending on `segmentation ` variable are added. Is it for removing unnecessary metric according to task?

File: src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py (hunk 13 above)
Comment (kprokofi) [reply src/otx/algorithms/common/adapters/mmcv/utils/_builder_build_data_parallel.py:160]: Yes, we don't need to put annotations on HPU if we compute loss on CPU. 
YOLOX from Habana team do the same -> they put post-processing on CPU. But I think it is not a good way to boost a perfomance, but other thing didn't work for OD

**Review by sungchul1 — APPROVED** [review]
LGTM. Please check pre-commit.

**Review by eunwoosh — APPROVED** [review]
